audit/audit.go

```diff
@@ -0,0 +1,272 @@
+}
+
+// CreateConn is a function that creates a connection to a remote audit server.
+type CreateConn func() (conn.Audit, error)
+
+// Option is an option for the smart client.
+type Option func(*Client) error
+
+// WithAuditOptions sets the audit options for the underlying audit client.
+func WithAuditOptions(options ...base.Option) Option {
+	return func(c *Client) error {
+		c.auditOptions = options
+		return nil
+	}
+}
+
+// WithBackoffPolicy sets the exponential backoff policy for making a new connection to the remote audit server.
+// By default this uses the default backoff settings from the exponential package.
+func WithExponentialBackoff(policy exponential.Policy) Option {
+	return func(c *Client) error {
+		c.policy = policy
+		return nil
+	}
+}
+
+// New creates a new smart client to the remote audit server. You can get a CreateConn
+// from conn.NewDomainSocket(), conn.NewTCP() or conn.NewNoop(). The last one is useful for testing
+// when you don't want to send logs anywhere.
+func New(create CreateConn, options ...Option) (*Client, error) {
+	connect, err := create()
+	if err != nil {
+		return nil, err
+	}
+
+	if runtime.GOOS != "linux" && !testing.Testing() {
+		if connect.Type() != conn.TypeNoOP {
+			return nil, fmt.Errorf("audit: only linux is clients can use Audit with non-NoOp connections")
+		}
+	}
+
+	c := &Client{
+		create:   create,
+		notifier: make(chan NotifyError, 1),
+	}
+	c.replaceBackoffRunner = c.replaceBackoff
+	c.replacingClient.Store(false)
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if reflect.ValueOf(c.policy).IsZero() {
+		var err error
+		c.backoff, err = exponential.New()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		c.backoff, err = exponential.New(exponential.WithPolicy(c.policy))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create the audit client.
+	aClient, err := base.New(connect, c.auditOptions...)
+	if err != nil {
+		return nil, err
+	}
+	c.client = aClient
+	c.sendRunner = c.client.Send
+
+	return c, nil
+}
+
+// Notify returns a channel that will receive errors when the connection to the remote audit server is broken or
+// the queue is full. If this channel is full, the error will be dropped.
+func (c *Client) Notify() <-chan NotifyError {
+	return c.notifier
+}
+
+// Send sends a message to the remote audit server. If the connection is broken, it will attempt to reconnect
+// but you will not receive an error. The only errors that will be returned are due to the Record being
+// invalid, trying to send a Msg with a type not DataPlane or ControlPlane or when we receive an
+// uncategorized error (which always indicates a handling bug in the Client).
+// Context timeouts will cause the write to drop, but not an error as all sends are
+// non-blocking. As a rule, you should not use Context timeouts as this is async. As a best practice,
+// you should do client.Send(context.WithoutCancel(ctx), msg) to prevent timeouts.
```

Comment:
Still seems weird we honor them at all, could just not check ctx cancellation in sync or async side here right?

Reply:
We don't, I removed that yesterday and I just removed this commet.

Reply:
Ok good that's more in line with what I thought you had said =)